download/download.go

```diff
@@ -114,13 +116,24 @@ func (d *Downloader) loop() {
 	var retry int
 
 	for {
-		err := d.oneShot(ctx)
+		longPoll, err := d.oneShot(ctx)
 		var delay time.Duration
 
 		if err == nil {
-			min := float64(*d.config.Polling.MinDelaySeconds)
-			max := float64(*d.config.Polling.MaxDelaySeconds)
-			delay = time.Duration(((max - min) * rand.Float64()) + min)
+			if !longPoll {
+				if d.config.Polling.LongPollingTimeoutSeconds != nil {
+					d.config.Polling.LongPollingTimeoutSeconds = nil
+				}
+
+				// revert the response header timeout value on the http client's transport
+				if *d.client.Config().ResponseHeaderTimeoutSeconds == 0 {
+					d.client = d.client.SetResponseHeaderTimeout(&d.respHdrTimeoutSec)
```

Comment:
I'm not certain, but since the client is passed in via [`New`](https://github.com/open-policy-agent/opa/pull/3418/files#diff-f808512360f39e5c4f67342ca24d84117be5403637b032d67e82df3a425cdde9R60), is it safe to change its transport settings? I'm wondering if it _could_ be used for other things, too.

Reply:
Performed a test which used _discovery_ with _periodic polling_ and the discovered config contained a _service bundle_ that used _long polling_. Both discovery and the service bundle are fetched from the same remote server.

<ins>Observation</ins>: The _response header timeout_ value on the http client’s transport is set to the **default** when the discovery bundle is being fetched and its set to **0**  when the service bundle is being fetched.

In the downloader we make a shallow copy of the client and then we update the timeout based on whether long polling is expected or not which seems to result in the correct behavior.

Reply:
:+1: thanks for checking this!